[TACHYON-255] Support for OpenStack Swift

gilv opened the pull request:
Support for OpenStack Swift and SoftLayer public object store
What is not here:
*********************
1. Documentation. Will submit it later in separate patch.
2. Testing is not here yet. Will be added later

How it works:
***************
First of all you need some local Swift or an account account in SoftLayer.
Those who like to setup local Swift - it's a very easy procedure ( see http://docs.openstack.org/developer/swift/development_saio.html )

My patch uses hadoop-openstack project, which internally implements Swift API and expose HDFS interface, so hadoop-openstack can easily be plugged into Tachyon.
It's a stable project, part of Hadoop releases from 2.4.0.  Hadoop community provides fixes and improves this driver constantly.

Few known issues:
***********************
Swift may come with various authentication models. 
It shipped with v1 authentication model, based on temp auth. However many uses v2 authentication model, based on Keystone. Current release of hadoop-openstack project has implementation based on Keystone. This means that your local Swift should be configured with Keystone.
SoftLayer doesn't use Keystone and use v1 auth model instead. The patch for the hadoop-driver to support v1 is not yet merged into Hadoop. 
Therefore those who wish to use SoftLayer object store need to follow more instructions ( until the pending patch will be merged into Hadoop)
The instructions are here https://gist.github.com/gilv/9d1e60aef201e76ffdba ( just follow Hadoop part section )

How to configure Tachyon
********************************
1. /conf folder contains core-site.xml.template
This file contains three sample configurations, one for Dallas, second local Swift with Keystone, third local Swift with temp auth.Depends on what you use, you need to configure them and add credentials and access URLs.

2. Rename core-site.xml.template file to core-site.xml and keep it in the same directory ( or not...it's configurable option in tachyon-end.sh or in 
tachyon.properties under tachyon.underfs.hadoop.configuration=${tachyon.home}/conf/core-site.xml

3. See tachyon-env.sh.swift how to run it. There is basically only one modification.

I tested that ./bin/tachyon runTest Basic CACHE_THROUGH works and also managed to sync existing objects in Swift with Tachyon.

AmplabJenkins commented:
Can one of the admins verify this patch?

haoyuan commented:
Jenkins, test this please.

rvesse commented on pom.xml:
```diff
@@ -124,7 +124,7 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <java.version>1.6</java.version>
-    <hadoop.version>1.0.4</hadoop.version>
+    <hadoop.version>2.6.0</hadoop.version>
```
While I personally would love Tachyon to target Hadoop 2.x by default I expect that you will get some pushback on this particular change

gilv replied:
Ohh...good point. I will remove this one.

rvesse commented on underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java:
```diff
@@ -66,9 +66,10 @@ public HdfsUnderFileSystem(String fsDefaultName, TachyonConf tachyonConf, Object
-
     prepareConfiguration(fsDefaultName, tachyonConf, tConf);
-
+    Utils.addKey(tConf, tachyonConf, "tachyon.underfs.hadoop.configuration");
```
Is there any reason for this not to go in the `prepareConfiguration()` method?

gilv replied:
I will move this line to prepareConfiguration() as you suggested.

calvinjia commented:
@gilv Thanks for the patch! Could you include the documentation in docs/Setup-UFS.md?

haoyuan commented:
Jenkins, add to whitelist.

haoyuan commented:
Jenkins, test this please.

haoyuan commented:
Jenkins, test this please.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/<h2>Failed Tests: <span class='status-failure'>4</span></h2><h3><a name='org.tachyonproject:tachyon' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon/testReport'>org.tachyonproject:tachyon</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon/testReport/tachyon.conf/TachyonConfTest/testMasterDefault/'><strong>tachyon.conf.TachyonConfTest.testMasterDefault</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon/testReport/tachyon.conf/TachyonConfTest/testCommonDefault/'><strong>tachyon.conf.TachyonConfTest.testCommonDefault</strong></a></li></ul><h3><a name='org.tachyonproject:tachyon-underfs-hdfs' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon-underfs-hdfs/testReport'>org.tachyonproject:tachyon-underfs-hdfs</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon-underfs-hdfs/testReport/tachyon.conf/TachyonConfTest/testMasterDefault/'><strong>tachyon.conf.TachyonConfTest.testMasterDefault</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3104/org.tachyonproject$tachyon-underfs-hdfs/testReport/tachyon.conf/TachyonConfTest/testCommonDefault/'><strong>tachyon.conf.TachyonConfTest.testCommonDefault</strong></a></li></ul>
Test FAILed.

gilv commented:
i will check why TachyonConfTest failed

gilv commented:
right, i see why it failed. i did some modifications to the properties, before i change it back - i would like to ask a question related it:

1. So the original value of tachyon.master.journal.folder was pointed to ${tachyon.home}/journal/.
It looks like to me that it's wrong...If someone use particular file system implementation - he would want to store all there and not to store journal in the home directory
2.  The same related tachyon.master.temporary.folder, i changed it instead of /tmp to ${tachyon.underfs.address}/tmp... 

Should i propose this as a separate patch or rollback completely - or you agree with me and i should modify tests?

-tachyon.master.journal.folder=${tachyon.home}/journal/
+#tachyon.master.journal.folder=${tachyon.home}/journal/
+tachyon.master.journal.folder=${tachyon.underfs.address}/journal/
tachyon.master.format.file_prefix=_format_
tachyon.master.port=19998
tachyon.master.web.port=19999
-tachyon.master.temporary.folder=/tmp
+tachyon.master.temporary.folder=${tachyon.underfs.address}/tmp

rvesse commented:
The journal is only relevant for the master hence why it is placed on the local file system.  Also the journal needs to be as fast to access as possible in order to avoid slowdowns because Tachyon uses the journal as a write ahead log so updates to the file system metadata are gated by the speed of the journal access.

The journal should only be placed on a shared file system if you are setting up fault tolerance per the [Fault Tolerant Tachyon Cluster](http://tachyon-project.org/Fault-Tolerant-Tachyon-Cluster.html) documentation in which case all the masters need to be able to access the journal.

Similarly the master temporary folder is just that - temporary - so again it makes sense for it to be local to the master since it doesn't need to be available anywhere else in the cluster and access to it wants to be as fast as possible

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3108/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='org.tachyonproject:tachyon' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3108/org.tachyonproject$tachyon/testReport'>org.tachyonproject:tachyon</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3108/org.tachyonproject$tachyon/testReport/tachyon.conf/TachyonConfTest/testCommonDefault/'><strong>tachyon.conf.TachyonConfTest.testCommonDefault</strong></a></li></ul><h3><a name='org.tachyonproject:tachyon-underfs-hdfs' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3108/org.tachyonproject$tachyon-underfs-hdfs/testReport'>org.tachyonproject:tachyon-underfs-hdfs</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3108/org.tachyonproject$tachyon-underfs-hdfs/testReport/tachyon.conf/TachyonConfTest/testCommonDefault/'><strong>tachyon.conf.TachyonConfTest.testCommonDefault</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3109/
Test PASSed.

calvinjia commented:
@gilv Could you update this with the latest master?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3134/<h2>Build result: FAILURE</span></h2>[...truncated 2049 lines...][ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :tachyon-underfs-hdfs[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/pom.xml to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.7.0-SNAPSHOT/tachyon-underfs-hdfs-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/pom.xml to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-tests.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/assembly/pom.xml to org.tachyonproject/tachyon-assemblies/0.7.0-SNAPSHOT/tachyon-assemblies-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.7.0-SNAPSHOT/tachyon-parent-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/client/pom.xml to org.tachyonproject/tachyon-client/0.7.0-SNAPSHOT/tachyon-client-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/pom.xml to org.tachyonproject/tachyon-underfs/0.7.0-SNAPSHOT/tachyon-underfs-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/examples/pom.xml to org.tachyonproject/tachyon-examples/0.7.0-SNAPSHOT/tachyon-examples-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.7.0-SNAPSHOT/tachyon-shell-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.7.0-SNAPSHOT/tachyon-underfs-glusterfs-0.7.0-SNAPSHOT.pomchannel stopped
Test FAILed.

gilv commented:
It failed since Hadoop-openstack is not exists in the Hadoop 1.X.X..
Is there way to modify pom.xml so it will be included only for Hadoop 2.4.0 and up? Or i should move it to comment?

[ERROR] Failed to execute goal on project tachyon-underfs-hdfs: Could not resolve dependencies for project org.tachyonproject:tachyon-underfs-hdfs:jar:0.7.0-SNAPSHOT: Could not find artifact org.apache.hadoop:hadoop-openstack:jar:1.0.4 in central (https://repo1.maven.org/maven2) -> [Help 1]

calvinjia commented:
@gilv Its my suspicion that lower versions of hadoop (ie. 1.x) will not work with the hadoop openstack client (2.x). Its worth a try to see if it can work though. In case it does, you can just use another variable to specify the hadoop openstack version since its no longer tied to the actual hadoop version. In the case it doesn't work, one solution would be to modularize it like glusterfs and we document that hadoop 2.x is required to use it.

gilv commented:
@calvinjia Hadoop-openstack requires version 2.4.0 and up, it will not work with versions 1.x.x
I will try to look how you resolved it for GlusterFS.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3143/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3144/
Test PASSed.

calvinjia commented on core/pom.xml:
```diff
@@ -96,8 +96,13 @@
-      <version>3.0.1</version>
-      <scope>test</scope>
+      <version>3.1</version>
+      <!--scope>test</scope-->
```
Is this required by hadoop-openstack?

gilv replied:
Yes, 
should i add it also dependent on hadoop-openstack,version?

calvinjia commented:
@gilv Could you modularize the swift integration into another module under `underfs` like `underfs/swift` (similar to `underfs/glusterfs`)? Although there is not too much code change in the hdfs underfilesystem, I think it would be cleaner to keep it separate from hdfs, especially if we add changes in the future.

gilv commented:
@calvinjia : But the point is to use HDFS interfaces, why would you like to have the same code twice? The hdfs module you have implements HDFS interfaces and than we can use it for Swift or S3 or HDFS... I think it's very good and make code cleaner, otherwise i will need to copy 99% of the code to /swift module...

calvinjia commented:
You should be able to extend the hdfs interface and override the prepare configuration portion? Or do you mean there's just too much boiler plate to write for a new submodule?

calvinjia commented:
@gilv I think the change is minor enough to bundle it into the hdfs implementation. In the future if there are more specific things that won't be encapsulated by the conf, we can think about moving to another submodule.

I have two other concerns:

1. I think the open stack dependency will bring in dependencies like hadoop-common and other libraries that Tachyon already has. We should manage these versions because if a user wishes to use hadoop 1.x, openstack 2.x will still be included and possibly bring in conflicting libraries. 

2. Could you add documentation to docs/Setup-UFS.md?

gilv commented:
@calvinjia : I will add documentation. sure.
Related your concern about dependencies: what is the reason Tachyon doesn't use Hadoop 2.4.0 by default (or up) for hdfs module? In any case it just some dependencies and no  one really need to install Hadoop for this, so why not to addapt hdfs module to use 2.4.0?

calvinjia commented:
@gilv 
Currently Tachyon's default version is 1.0.4, but I can see an argument for using 2.x. But if the user wishes to use Tachyon with Hadoop 1.x, we should still support that. Bringing in 2.x hadoop dependencies in the hdfs module (from open stack, since its included regardless of if the user is intending to use it) will probably cause issues when the 1.x hadoop client is used.

gilv commented:
@calvinjia, i also think there will be issues with Hadoop 1.x client. Does GlusterFS has the same issues?  Is there a way to modify pom file so that it will use Hadoop-openstack only when Tachyon compiled with Hadoop 2,4.0 and up?

calvinjia commented:
@gilv Actually since the code doesn't use anything from hadoop-openstack directly, you could just include it in a profile so by default it would not be included. If the user wants to use swift they would need to rebuild Tachyon like `mvn clean install -Pswift` if the profile was called swift.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3167/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3179/
Test PASSed.

calvinjia commented:
@gilv could you rebase/merge with master again? Seems like there are some conflicts.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3181/<h2>Build result: FAILURE</span></h2>[...truncated 1814 lines...][ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :tachyon-underfs-hdfs[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/pom.xml to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/common/target/tachyon-common-0.7.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-common/0.7.0-SNAPSHOT/tachyon-common-0.7.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.7.0-SNAPSHOT/tachyon-underfs-hdfs-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/pom.xml to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/core/target/tachyon-0.7.0-SNAPSHOT-tests.jar to org.tachyonproject/tachyon/0.7.0-SNAPSHOT/tachyon-0.7.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/assembly/pom.xml to org.tachyonproject/tachyon-assemblies/0.7.0-SNAPSHOT/tachyon-assemblies-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/integration-tests/pom.xml to org.tachyonproject/tachyon-integration-tests/0.7.0-SNAPSHOT/tachyon-integration-tests-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.7.0-SNAPSHOT/tachyon-parent-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/client/pom.xml to org.tachyonproject/tachyon-client/0.7.0-SNAPSHOT/tachyon-client-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/pom.xml to org.tachyonproject/tachyon-underfs/0.7.0-SNAPSHOT/tachyon-underfs-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/examples/pom.xml to org.tachyonproject/tachyon-examples/0.7.0-SNAPSHOT/tachyon-examples-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/s3/pom.xml to org.tachyonproject/tachyon-underfs-s3/0.7.0-SNAPSHOT/tachyon-underfs-s3-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.7.0-SNAPSHOT/tachyon-shell-0.7.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.7.0-SNAPSHOT/tachyon-underfs-glusterfs-0.7.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3182/
Test PASSed.

calvinjia commented on core/pom.xml:
```diff
@@ -119,6 +119,70 @@
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
```
You should be able to add a simpler profile to `underfs/hdfs/pom.xml` just including the open-stack dependency.

gilv commented:
this also doesn't work for me now...something is wrong after latest refactoring, trying to debug it.

gilv commented:
@calvinjia Do you have any idea why i am getting this? 

org.apache.hadoop.fs.FileSystem: Provider tachyon.hadoop.TFS not found

	at tachyon.underfs.UnderFileSystemRegistry.create(UnderFileSystemRegistry.java:132)
	at tachyon.underfs.UnderFileSystem.get(UnderFileSystem.java:85)
	at tachyon.underfs.UnderFileSystem.get(UnderFileSystem.java:69)
	at tachyon.Format.formatFolder(Format.java:38)
	at tachyon.Format.main(Format.java:66)

calvinjia commented:
@gilv have you synced to the latest master?

gilv commented:
@calvinjia the last time i updated was two days ago and got those exceptions and was sure it come from my code... I just updated as you suggested and they are gone. What was that exception? I tried to debug it and couldn't understand the origin of it

calvinjia commented:
@gilv That exception probably occurred because the hadoop FileSystem meta-inf file was in the wrong place (stating that an implementation, `tachyon.hadoop.TFS` exists in the module but really doesn't). Lots of files were moved around in the past few days and its possible that there was a bad state at some time.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3199/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.tachyonproject:tachyon-common' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3199/org.tachyonproject$tachyon-common/testReport'>org.tachyonproject:tachyon-common</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3199/org.tachyonproject$tachyon-common/testReport/tachyon.conf/TachyonConfTest/testCommonDefault/'><strong>tachyon.conf.TachyonConfTest.testCommonDefault</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3200/
Test PASSed.

gilv commented:
@calvinjia : I saw the recent structure of underfs module. I am 100% agree now with your previous suggestion to make a separate module for Swift. I think it's much cleaner and can be better supported for specific Swift needs.

calvinjia commented:
@gilv 
Thanks for making a new submodule!

Instead of extending `UnderFileSystem`, it may be cleaner to extend `HdfsUnderFileSystem` and override only the methods that require special changes for swift.

Could you also take a look at a underfs submodule pom (ie. `underfs/hdfs/pom.xml`) and see how we are structuring the dependencies after the work of splitting up the core jar?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3202/
Test PASSed.

gilv commented:
@calvinjia Good points. Please review my changes.

calvinjia commented:
@gilv Looks good overall. I think you can revert the changes made to `core/pom.xml`?

gilv commented:
@calvinjia I am not sure why, but without modification to core/pom.xml - it doesn't work.

gilv commented:
@calvinjia : you mean to remove the swift profile from core/pom.xml? it doesn't work without it, Not sure why

calvinjia commented:
@gilv I actually mean all the changes done to `core/pom.xml`. What kind of error do you get when it is reverted?

gilv commented:
@calvinjia  If there are no further remarks, can we merge it?. Changes to core/pom.xml will be in separate patch? Thanks

calvinjia commented:
@gilv After merging master you should be able to revert the changes in `core/pom.xml` and everything should still work. You will need to add the new swift module to `assembly/pom.xml`.

gilv commented:
@calvinjia not sure why it happens, but the some of merge was manual and than it also included my changes to core/pom.xml. This is why this commit has all the changes from master...

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3211/
Test PASSed.

gilv commented:
is there anything else? Can we merge it?

calvinjia commented:
@gilv it seems like the changes to `core/pom.xml` were not reverted? Or did you mean that you couldn't due to some git weirdness?

gilv commented:
@calvinjia  I did changes to it, i reverted all changes. will check it again now

gilv commented:
i don't see core module anymore...you make changes to rapidly :)

calvinjia commented:
@gilv it's renamed to tachyon-servers now

gilv commented:
@calvinjia  so i don't have any modification to servers/pom.xml, there is no single word about openstack in this file, what changes you mean i  need to revert there?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3223/
Test PASSed.

haoyuan commented on underfs/swift/src/test/java/tachyon/underfs/swift/SwiftUnderFileSystemTest.java:
```diff
@@ -0,0 +1,35 @@
+    TachyonConf conf = new TachyonConf();
+    
+    UnderFileSystemFactory factory =  UnderFileSystemRegistry.find("swift://localhost/test/path", conf);
+    Assert.assertNotNull("A UnderFileSystemFactory should exist for Swift paths when using this module", factory);
```
much longer than 100 chars. 

why doesn't the style checker check this?

haoyuan commented on underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystemFactory.java:
```diff
@@ -0,0 +1,44 @@
+      return true;
+    }
+    return false;
+    
```
extra empty line.

haoyuan commented on underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystem.java:
```diff
@@ -0,0 +1,62 @@
+public class SwiftUnderFileSystem extends HdfsUnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+
```
extra line.

haoyuan commented on servers/pom.xml:
```diff
@@ -94,13 +94,12 @@
-      <version>3.0.1</version>
-      <scope>test</scope>
+      <version>3.1</version>
+      <!--scope>test</scope-->
```
why was this changed?

calvinjia replied:
Let's change this back to test and then include it in the swift pom.

calvinjia commented on servers/pom.xml:
```diff
@@ -119,6 +118,51 @@
+      </dependencies>
+    </profile>
+
+  </profiles>
```
These changes shouldn't be needed.

calvinjia commented on underfs/swift/src/test/resources/test_metrics.properties:
```diff
@@ -0,0 +1,6 @@
+*.source.jvm.class=tachyon.metrics.source.JvmSource
+master.sink.console.period=20
+master.sink.console.unit=minutes
+master.sink.jmx.class=tachyon.metrics.sink.JmxSink
```
This file is no longer necessary.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3225/
Test PASSed.

gilv commented:
@haoyuan : I run mvn checkstyle on swift module

[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Tachyon Under File System - Swift 0.7.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:2.13:checkstyle (default-cli) @ tachyon-underfs-swift ---
[INFO] Starting audit...
Audit done.

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.309s
[INFO] Finished at: Fri May 15 11:21:47 IDT 2015
[INFO] Final Memory: 24M/217M
[INFO] ------------------------------------------------------------------------

calvinjia commented:
A few minor formatting issues, but overall LGTM! Thanks for doing this PR despite the refactoring of the overall codebase, its a very important feature.

haoyuan commented:
Thanks for the great feature!
